pytorch_lightning/strategies/ddp.py

```diff
@@ -421,7 +422,11 @@ def teardown(self) -> None:
         if isinstance(self.model, DistributedDataParallel):
             self.model = self.lightning_module
 
-        if self._layer_sync:
+        if (
+            self.lightning_module.trainer is not None
```

Comment:
Do we need to check this? Trainer won't be none right?

Reply:
The trainer is expected to exist there unless users temper with their lightning_module.

Reply:
The tests fail otherwise:

```
>       if self.lightning_module.trainer.state.fn == TrainerFn.FITTING and self._layer_sync:
E       AttributeError: 'NoneType' object has no attribute 'state'
```

I'll look into this closer.

Reply:
This was discussed here: https://github.com/PyTorchLightning/pytorch-lightning/pull/11919#discussion_r810444367

Reply:
> The trainer is expected to exist there unless users temper with their lightning_module.

lightning_module.trainer is None on the main process in ddp_spawn

A None check is required for exactly this reason. It was removed in the latest commit however. Let's not go in circles 😄 Let's add a comment there why we do the None check and be done with it so we can merge the PR.